src/mqueue.rs

```diff
@@ -4,142 +4,126 @@
-#[cfg(target_os = "linux")]
-mod consts {
-    use libc::c_int;
-
-    bitflags!(
-        flags MQ_OFlag: c_int {
-            const O_RDONLY    = 0o00000000,
-            const O_WRONLY    = 0o00000001,
-            const O_RDWR      = 0o00000002,
-            const O_CREAT     = 0o00000100,
-            const O_EXCL      = 0o00000200,
-            const O_NONBLOCK  = 0o00004000,
-            const O_CLOEXEC   = 0o02000000,
-        }
-    );
-
-    bitflags!(
-        flags FdFlag: c_int {
-            const FD_CLOEXEC = 1
-        }
-    );
+use std::mem;
+
+bitflags!{
+    flags MQ_OFlag: libc::c_int {
+        const O_RDONLY    = 0o00000000,
+        const O_WRONLY    = 0o00000001,
+        const O_RDWR      = 0o00000002,
+        const O_CREAT     = 0o00000100,
+        const O_EXCL      = 0o00000200,
+        const O_NONBLOCK  = 0o00004000,
+        const O_CLOEXEC   = 0o02000000,
+    }
 }
 
-mod ffi {
-    use libc::{c_char, size_t, ssize_t, c_uint, c_int};
-    use super::MQd;
-    use super::MqAttr;
-
-    #[allow(improper_ctypes)]
-    extern "C" {
-        pub fn mq_open(name: *const c_char, oflag: c_int, ...) -> MQd;
-
-        pub fn mq_close (mqd: MQd) -> c_int;
-
-        pub fn mq_unlink(name: *const c_char) -> c_int;
-
-        pub fn mq_receive (mqd: MQd, msg_ptr: *const c_char, msg_len: size_t, msq_prio: *const c_uint) -> ssize_t;
-
-        pub fn mq_send (mqd: MQd, msg_ptr: *const c_char, msg_len: size_t, msq_prio: c_uint) -> c_int;
-
-        pub fn mq_getattr(mqd: MQd, attr: *mut MqAttr) -> c_int;
-
-        pub fn mq_setattr(mqd: MQd, newattr: *const MqAttr, oldattr: *mut MqAttr) -> c_int;
+bitflags!{
+    flags FdFlag: libc::c_int {
+        const FD_CLOEXEC = 1
     }
 }
 
 #[repr(C)]
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[derive(Clone, Copy)]
 pub struct MqAttr {
-    pub mq_flags: c_long,
-    pub mq_maxmsg: c_long,
-    pub mq_msgsize: c_long,
-    pub mq_curmsgs: c_long,
-    pad: [c_long; 4]
+    mq_attr: libc::mq_attr
 }
 
-impl MqAttr {
-   pub fn new(mq_flags: c_long, mq_maxmsg: c_long, mq_msgsize: c_long, mq_curmsgs: c_long) -> MqAttr {
-       MqAttr { mq_flags: mq_flags, mq_maxmsg: mq_maxmsg, mq_msgsize: mq_msgsize, mq_curmsgs: mq_curmsgs, pad: [0; 4] }
-   }
+impl PartialEq<MqAttr> for MqAttr {
+    fn eq(&self, other: &MqAttr) -> bool {
+        let self_attr = self.mq_attr;
+        let other_attr = other.mq_attr;
+        self_attr.mq_flags == other_attr.mq_flags &&
+            self_attr.mq_maxmsg == other_attr.mq_maxmsg &&
+            self_attr.mq_msgsize == other_attr.mq_msgsize &&
+            self_attr.mq_curmsgs == other_attr.mq_curmsgs
+    }
 }
 
+impl MqAttr {
+    pub fn new(mq_flags: c_long, mq_maxmsg: c_long, mq_msgsize: c_long, mq_curmsgs: c_long) -> MqAttr {
+        let mut attr = unsafe { mem::uninitialized::<libc::mq_attr>() };
+        attr.mq_flags = mq_flags;
+        attr.mq_maxmsg = mq_maxmsg;
+        attr.mq_msgsize = mq_msgsize;
+        attr.mq_curmsgs = mq_curmsgs;
```

Comment:
Is there any reason to not use the struct initializer syntax in this case rather than creating a mutable `mq_attr` and setting every field?